modules/secure-preview/rest/secure-preview.php

```diff
@@ -0,0 +1,160 @@
+	/**
+	 * Initialize the class
+	 */
+	public static function init() {
+		add_action( 'rest_api_init', [ __CLASS__, 'register_routes' ] );
+	}
+
+	/**
+	 * Register the REST routes
+	 */
+	public static function register_routes() {
+		register_rest_route( VIP_WORKFLOW_REST_NAMESPACE, '/preview/(?P<post_id>[0-9]+)', [
+			'methods'             => 'POST',
+			'callback'            => [ __CLASS__, 'handle_generate_preview_token' ],
+			'permission_callback' => [ __CLASS__, 'permission_callback' ],
+			'args'                => [
+				// URL Parameters
+				'post_id'         => [
+					'required'          => true,
+					'validate_callback' => function ( $param ) {
+						$post_id = absint( $param );
+						return get_post( $post_id ) instanceof \WP_Post;
+					},
+					'sanitize_callback' => function ( $param ) {
+						return absint( $param );
+					},
+				],
+
+				// POST data parameters
+				'expiration'      => [
+					'required'          => true,
+					'validate_callback' => function ( $param ) {
+						$expiration_options = VIP_Workflow::instance()->secure_preview->get_link_expiration_options();
+						$expiration_values  = wp_list_pluck( $expiration_options, 'value' );
+
+						return in_array( $param, $expiration_values );
+					},
+					'sanitize_callback' => function ( $param ) {
+						return strval( $param );
+					},
+				],
+				'is_one_time_use' => [
+					'required'          => true,
+					'validate_callback' => function ( $param ) {
+						return true === $param || false === $param;
+					},
+					'sanitize_callback' => function ( $param ) {
+						return boolval( $param );
+					},
+				],
+
+			],
+		] );
+	}
+
+	/**
+	 * Users allowed to edit the post can generate preview links
+	 */
+	public static function permission_callback( WP_REST_Request $request ) {
+		$post_id = $request->get_param( 'post_id' );
+		return current_user_can( 'edit_posts', $post_id );
+	}
+
+	/**
+	 * Handle a request to create a preview token
+	 *
+	 * @param WP_REST_Request $request
+	 */
+	public static function handle_generate_preview_token( WP_REST_Request $request ) {
+		$post_id          = $request->get_param( 'post_id' );
+		$expiration_value = $request->get_param( 'expiration' );
+		$is_one_time_use  = $request->get_param( 'is_one_time_use' );
+
+		if ( ! VIP_Workflow::instance()->custom_status->is_post_using_custom_status( $post_id ) ) {
+			$post_status = get_post_status( $post_id );
+
+			if ( 'publish' === $post_status ) {
+				return new WP_Error( 'vip-workflow-published-post', __( 'Secure links can not be generated for published posts.', 'vip-workflow' ) );
+			} elseif ( 'auto-draft' === $post_status ) {
+				return new WP_Error( 'vip-workflow-not-custom-status', __( 'Posts must be saved before a secure link can be generated.', 'vip-workflow' ) );
+			} else {
+				return new WP_Error( 'vip-workflow-not-custom-status', __( 'Secure links can only be generated for pre-published posts with a custom status.', 'vip-workflow' ) );
+			}
+		}
```

Comment:
Given this validation when the token is created, shouldn't the post status be an input to the token envelope and be part of the token validation? In other words, if the post status changes (e.g., the post is published), shouldn't the token be invalidated?

Reply:
Would echo the same Q

Reply:
> Given this validation when the token is created, shouldn't the post status be an input to the token envelope and be part of the token validation?

Since we have multiple pre-publish statuses, I'm not sure we want to tie a token to a particular custom status. Let's say we're working with a set of custom statuses like this:

1. Pitch
2. Draft
3. Pending review

It would seem valid to me to create a token for "Pitch" and give it to a colleague, and then maybe promote the post to "Draft" in the meantime. Or the same between "Draft" and "Pending review".

I think the more important part is that a token made for a post in a custom status should only be valid while that post has a custom status assigned to it. What do you think about changing `validate_token()` to ensure that the token is only valid for a post using a custom status, and any other statuses are invalid? This gives some flexibility to modify tracking statuses in a workflow without invalidating links, but ensures they only apply to our custom post statuses.

Reply:
i get what you mean, they are intending for one link to be used throughout the process. In the event that they don't want that, then a one time use token could work or the max time could be shortened.

That said, I'm wondering if this makes sense as a setting that customers could toggle, in case they are conscious of security and want a new token. That way, by default it won't get invalidated but if they want to make it stricter they can. This would be similar to the [cascading/restrictive hierarchy filter we have in the governance plugin](https://github.com/Automattic/vip-governance-plugin?tab=readme-ov-file#vip_governance__is_block_allowed_in_hierarchy).

Reply:
I added some code to explicitly ignore and clear tokens when a preview for a non-custom-status or non-supported post type is used:

https://github.com/Automattic/vip-workflow-plugin/blob/84e5bc42e46db0926a32652a206a47f8d5d0a37f/modules/preview/token.php#L57-L64

Here's how that will work with some different situations:

1. If a preview URL is generated for a post with custom status "Pitch", and then it moves to another custom status "Draft", the key will continue to work.
2. If a post has a token URL generated, and then moves to a non-custom status (either "Publish" or another custom status that isn't managed by our plugin), the key will be rejected and all tokens will be removed.
3. If a post has a token URL generated in a custom status, then briefly moves to "Publish", and then moves back to a custom statues (e.g. "Pending Review"), the URL will continue to work, as long as it wasn't used while the post was in the "Publish" state.

The third situation above may seem a little odd, but I think it's not harmful if this situation occurs. I opted to do key removal in the `validate_token()` function rather than relying on an external `transition_post_status` or similar filter to remove tokens automatically in the background. The main reasoning is that this simplifies the logic - we only need to think about token management in one function, and there are no other side-effects that will also manage tokens. Also, just in case a post doesn't fire the expected hooks (like if a post was updated to `publish` directly in the database), we'll still need this logic in `validate_token()` anyhow.

This solution keeps "security"-related logic in a single flow in one function, which I think is worthwhile.